File: FSharp/EntityFramework/Context/ApplicationContextFactory.fs
namespace Context

open Microsoft.EntityFrameworkCore
open Microsoft.EntityFrameworkCore.Design

type ApplicationContextFactory (_connectionString) as this =
  interface IDesignTimeDbContextFactory<ApplicationContext> with
    member _.CreateDbContext parameters =
      if Array.length parameters > 0 then this.GetContext parameters[1] else this.GetContext ()

  new() = ApplicationContextFactory "Filename=db.bd"

  member _.GetContext (?connectionString: string) =
    let mutable builder = new DbContextOptionsBuilder<ApplicationContext> ()
    builder <- builder.UseSqlite(if connectionString.IsSome then connectionString.Value else _connectionString)
    builder <- builder.EnableSensitiveDataLogging ()

    let ctx = new ApplicationContext (builder.Options)
    ctx.Init ()
    ctx

File: CSharp/EntityFramework/Context/ApplicationContextFactory.cs
namespace Context;

using Microsoft.EntityFrameworkCore;
using Microsoft.EntityFrameworkCore.Design;

public class ApplicationContextFactory : IDesignTimeDbContextFactory<ApplicationContext>
{
    private readonly string _connectionString;

    public ApplicationContextFactory()
    {
        _connectionString = "Filename=db.bd";
    }

    public ApplicationContextFactory(string connectionString)
    {
        _connectionString = connectionString;
    }

    public ApplicationContext GetContext(string? connectionString = null)
    {
        var builder = new DbContextOptionsBuilder<ApplicationContext>()
            .UseSqlite(connectionString ?? _connectionString)
            .EnableSensitiveDataLogging();

        var ctx = new ApplicationContext(builder.Options);
        return ctx;
    }

    public ApplicationContext CreateDbContext(string[] parameters) =>
        GetContext(parameters.Count() > 0 ? parameters[0] : null);
}
